### PR TITLE
Enable nullable reference types for Microsoft.Build.Tasks.CodeAnalysis

### PR DESCRIPTION
**Pull request by sharwell**
Recommended review strategy is commit-by-commit

### CODE REVIEW

File: src/Compilers/Core/MSBuildTask/CanonicalError.cs [hunk 1]
```diff
@@ -1,5 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+#nullable enable
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/CanonicalError.cs]: Can we enable nullability at the project-level instead?

File: src/Compilers/Core/MSBuildTask/CanonicalError.cs [hunk 2]
```diff
@@ -166,7 +168,7 @@ internal Parts()
             /// <summary>
             /// Name of the file or tool (not localized)
             /// </summary>
-            internal string origin;
+            internal string? origin;
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/CanonicalError.cs]: >origin [](start = 29, length = 6)

How do we get `null` in this field?

**Comment by jcouv** [comment]
>         internal static bool IsParameterEmpty(string parameter, params char[] splitOn)

?

---
Refers to: src/Compilers/Core/MSBuildTask/CommandLineBuilderExtension.cs:175 in 7dbd3c0. [](commit_id = 7dbd3c0231abe83307e74d1ac9e7769066b12fb4, deletion_comment = False)

File: src/Compilers/Core/MSBuildTask/CommandLineBuilderExtension.cs [hunk 3]
```diff
@@ -197,9 +199,9 @@ internal static bool IsParameterEmpty(string parameter, params char[] splitOn)
         internal void AppendSwitchIfNotNull
         (
             string switchName,
-            ITaskItem[] parameters,
+            ITaskItem[]? parameters,
             string[] metadataNames,
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/CommandLineBuilderExtension.cs]: >metadataNames [](start = 21, length = 13)

?

File: src/Compilers/Core/MSBuildTask/CanonicalError.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/MSBuildTask/CanonicalError.cs]: This would break tools used to evaluate the status of the Roslyn migration as a whole.

File: src/Compilers/Core/MSBuildTask/InteractiveCompiler.cs [hunk 4]
```diff
@@ -259,7 +261,7 @@ protected virtual void AddResponseFileCommands(CommandLineBuilderExtension comma
                 }
             }
 
-            if (ResponseFiles != null)
+            if (ScriptResponseFiles != null)
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/InteractiveCompiler.cs]: >ScriptResponseFiles [](start = 16, length = 19)

thanks for catching this. Will probably need a test for semantic change

File: src/Compilers/Core/MSBuildTask/CanonicalError.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/Core/MSBuildTask/CanonicalError.cs]: `new Parts()`

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs [hunk 5]
```diff
@@ -1035,8 +1037,11 @@ protected override bool HandleTaskExecutionErrors()
         /// <summary>
         /// Takes a list of files and returns the normalized locations of these files
         /// </summary>
-        private void NormalizePaths(ITaskItem[] taskItems)
+        private void NormalizePaths(ITaskItem[]? taskItems)
         {
+            if (taskItems is null)
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs]: >if (taskItems is null) [](start = 12, length = 22)

semantic change. Can this be hit?

File: src/Compilers/Core/MSBuildTask/Utilities.cs [hunk 6]
```diff
@@ -148,20 +150,14 @@ internal static void DeleteNoThrow(string path)
             return new ArgumentException(string.Format(CultureInfo.CurrentCulture, errorString, args));
         }
 
-        internal static string TryGetAssemblyPath(Assembly assembly)
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/Utilities.cs]: >TryGetAssemblyPath [](start = 31, length = 18)

is this semantic change necessary? Can it be moved to a separate PR?

File: src/Compilers/Core/MSBuildTaskTests/IntegrationTests.cs [hunk 7]
```diff
@@ -615,6 +622,8 @@ public void ReportAnalyzerMSBuild()
         [Fact(Skip = "failing msbuild")]
         public void SolutionWithPunctuation()
         {
+            RoslynDebug.Assert(s_msbuildExecutable is object);
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTaskTests/IntegrationTests.cs]: >s_msbuildExecutable [](start = 31, length = 19)

I'm wondering whether we could strengthen the type of `s_msbuildExecutable` to be non-nullable, by throwing from the static constructor

File: src/Compilers/Core/MSBuildTask/Utilities.cs (hunk 6 above)
Comment (sharwell) [reply src/Compilers/Core/MSBuildTask/Utilities.cs]: It's already a separate commit, but I can move it to a prerequisite PR if that's not enough.

File: src/Compilers/Core/MSBuildTaskTests/IntegrationTests.cs (hunk 7 above)
Comment (sharwell) [reply src/Compilers/Core/MSBuildTaskTests/IntegrationTests.cs]: I was able to strengthen it by making it an instance field.

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs (hunk 5 above)
Comment (sharwell) [reply src/Compilers/Core/MSBuildTask/ManagedCompiler.cs]: I have no idea

File: src/Compilers/Core/MSBuildTask/InteractiveCompiler.cs (hunk 4 above)
Comment (sharwell) [reply src/Compilers/Core/MSBuildTask/InteractiveCompiler.cs]: Where would this test go? I can't find any tests for `ResponseFiles` or `ScriptResponseFiles`.

File: src/Compilers/Core/MSBuildTask/CanonicalError.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/Core/MSBuildTask/CanonicalError.cs]: Consider doing both (`#nullable enable` in files, and also `Nullable=enable` in project). Enabling at the project level ensures that we will not regress as we add new files.

File: src/Compilers/Core/MSBuildTask/InteractiveCompiler.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/Core/MSBuildTask/InteractiveCompiler.cs]: Hum, doesn't appear to be any. The change is simple enough that I would not push for manual testing. Also the fact that this bug did not surface until now...

File: src/Compilers/Core/MSBuildTask/Utilities.cs (hunk 6 above)
Comment (jcouv) [reply src/Compilers/Core/MSBuildTask/Utilities.cs]: Sounds like it was not necessary. I'd remove that change from this PR, or request some manual validation, otherwise, the change makes me a bit nervous (not exercised).

File: src/Compilers/Core/MSBuildTask/CanonicalError.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/MSBuildTask/CanonicalError.cs]: @jaredpar any objection to this plan?

File: src/Compilers/Core/MSBuildTask/Utilities.cs (hunk 6 above)
Comment (sharwell) [reply src/Compilers/Core/MSBuildTask/Utilities.cs]: The reflection APIs return nulls.

**Comment by sharwell** [comment]
@jcouv @agocke I did a straightforward rebase of this now that #41180 was merged separately. I then implemented the suggestion from https://github.com/dotnet/roslyn/pull/40754#discussion_r369814126 which produce a single commit dbf6a03 that needs to be reviewed.

File: src/Compilers/Core/MSBuildTask/CanonicalError.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/MSBuildTask/CanonicalError.cs]: Now implemented

**Comment by sharwell** [comment]
@jcouv Can you confirm the updated commit mentioned in https://github.com/dotnet/roslyn/pull/40754#issuecomment-577933242?
